apps/interop-rxjs-signal/src/app/list/photos.component.ts

```diff
@@ -32,88 +46,122 @@ import { PhotoStore } from './photos.store';
-            >
-          </button>
-          Page :{{ vm.page }} / {{ vm.pages }}
-        </section>
-        <mat-progress-bar
-          mode="query"
-          *ngIf="vm.loading"
-          class="mt-5"></mat-progress-bar>
-        <ul
-          class="flex flex-wrap gap-4"
-          *ngIf="vm.photos && vm.photos.length > 0; else noPhoto">
-          <li *ngFor="let photo of vm.photos; trackBy: trackById">
-            <a routerLink="detail" [queryParams]="{ photo: encode(photo) }">
-              <img
-                src="{{ photo.url_q }}"
-                alt="{{ photo.title }}"
-                class="image" />
-            </a>
-          </li>
-        </ul>
-        <ng-template #noPhoto>
-          <div>No Photos found. Type a search word.</div>
-        </ng-template>
-        <footer class="text-red-500">
-          {{ vm.error }}
-        </footer>
+    <section class="flex flex-col">
+      <section class="flex gap-3 items-center">
+        <button
+          [disabled]="page() === 1"
+          [class.bg-gray-400]="page() === 1"
+          class="text-xl border rounded-md p-3"
+          (click)="previousPage()">
+          <
+        </button>
+        <button
+          [disabled]="endOfPage()"
+          [class.bg-gray-400]="endOfPage()"
+          class="text-xl border rounded-md p-3"
+          (click)="nextPage()">
+          >
+        </button>
+        Page :{{ page() }} / {{ pages() }}
       </section>
-    </ng-container>
+      <mat-progress-bar
+        mode="query"
+        *ngIf="loading()"
+        class="mt-5"></mat-progress-bar>
+      <ul
+        class="flex flex-wrap gap-4"
+        *ngIf="photos() && photos().length > 0; else noPhoto">
+        <li *ngFor="let photo of photos(); trackBy: trackById">
+          <a routerLink="detail" [queryParams]="{ photo: encode(photo) }">
+            <img
+              src="{{ photo.url_q }}"
+              alt="{{ photo.title }}"
+              class="image" />
+          </a>
+        </li>
+      </ul>
+      <ng-template #noPhoto>
+        <div>No Photos found. Type a search word.</div>
+      </ng-template>
+      <footer class="text-red-500">
+        {{ error() }}
+      </footer>
+    </section>
   `,
-  providers: [provideComponentStore(PhotoStore)],
   host: {
     class: 'p-5 block',
   },
 })
-export default class PhotosComponent implements OnInit {
-  store = inject(PhotoStore);
-  readonly vm$ = this.store.vm$.pipe(
-    tap(({ search }) => {
-      if (!this.formInit) {
-        this.search.setValue(search);
-        this.formInit = true;
-      }
-    })
+export default class PhotosComponent {
+  private photoService = inject(PhotoService);
+
+  searchForm = new FormControl();
+
+  readonly page = signal(1);
+  readonly search = signal('');
+  readonly loading = signal(false);
+  readonly error = signal('');
+  readonly endOfPage = computed(() => this.page() >= this.pages());
+
+  readonly photosAPI = computedFrom(
+    [this.search, this.page],
```

Comment:
```suggestion
    [this.searchForm.valueChanges, this.page],
```

Reply:
Or what we can do is to extract the valueChanges to it's own variable with filtering, debouncing and other stuff. 

```ts
searchFormChanges$ =  this.searchForm.valueChanges.pipe(
  filter(([search]) => search.length >= 3),
  debounceTime(300),
  distinctUntilChanged(),
);
```

and use it as: 

```ts
readonly photosAPI = computedFrom(
  [this.searchFormChanges$, this.page],
  pipe(
    tap(() => this.loading.set(true)),
    switchMap(([search, page]) =>
      this.photoService.searchPublicPhotos(search, page).pipe(
        catchError((err: unknown) => {
          this.error.set(JSON.stringify(err));
          return EMPTY;
        }),
        finalize(() => this.loading.set(false))
      )
    ),
    startWith({ photos: { photo: [] as Photo[], pages: 0 } })
  )
);
```

This way we don't have to debounce the api call if the page changes, but we want it only when doing the search in the input.

Reply:
yes I thought of it during lunch break. I didn't like the subscribe and the distinctUntilChange with the custom function. 

And you are right about the debounceTime, it's not needed for page change.

Thanks for the review and nice work on the computedFrom. 👍

Reply:
I still need the search property as a signal because i listen to it inside my effect to save it inside the localStorage, but using `toSignal` is still nicer than using a subscribe function.